calico_containers/tests/unit/profile_test.py

```diff
@@ -0,0 +1,432 @@
+            else:
+                self.assertEqual(m_Rules.outbound_rules, [rule, new_rule])
+
+    @parameterized.expand([
+        ('inbound', 0), ('inbound', 1), ('inbound', -5),
+        ('outbound', 0), ('outbound', 1), ('outbound', -5),
+    ])
+    def test_profile_rule_add_remove_add_rule_front_of_list(
+            self, direction_arg, position_arg):
+        """
+        Test for profile_rule_add_remove function when adding a new Rule to the
+        front of the inbound or outbound rules list.
+
+        Test for both directions - inbound and outbound.
+
+        Test for multiple positions less than 0
+        (including positions out of range)
+        """
+        with patch('calico_ctl.profile.client', autospec=True) as m_client:
+            # Setup arguments to pass to method under test
+            operation = 'add'
+            name = 'profile1'
+            position = position_arg
+            action = 'allow'
+            direction = direction_arg
+
+            # Set up Mock objects
+            rule = Rule()
+            m_Rules = Mock(spec=Rules, id=name, inbound_rules=[rule],
+                       outbound_rules=[rule])
+            m_Profile = Mock(spec=Profile, name=name, rules=m_Rules)
+            m_client.get_profile.return_value = m_Profile
+
+            # Set up new rule that function will create/add - compare in asserts
+            rule_dict = {'action': 'allow', 'icmp_type':5, 'icmp_code':5}
+            new_rule = Rule(**rule_dict)
+
+            # Call method under test
+            profile_rule_add_remove(operation, name, position, action, direction,
+                                    icmp_type='5', icmp_code='5')
+
+            # Assert
+            m_client.get_profile.assert_called_once_with(name)
+            m_client.profile_update_rules.assert_called_once_with(m_Profile)
+            if direction_arg=='inbound':
+                self.assertEqual(m_Rules.inbound_rules, [new_rule, rule])
+            else:
+                self.assertEqual(m_Rules.outbound_rules, [new_rule, rule])
+
+    def test_profile_rule_add_remove_invalid_protocol(self):
+        """
+        Test for profile_rule_add_remove when passing a protocol (besides udp
+        or tcp) and src/dst ports.
+
+        Assert that the system exits
+        """
+        # Setup arguments to pass to method under test
+        operation = 'add'
+        name = 'profile1'
+        position = None
+        action = 'allow'
+        direction = 'inbound'
+
+        # Call method under test
+        self.assertRaises(SystemExit, profile_rule_add_remove,
+                          operation, name, position, action, direction,
+                          protocol='icmp', src_ports=[40,60])
+
+    @patch('calico_ctl.profile.client', autospec=True)
+    def test_profile_rule_add_remove_fail_get_profile(self, m_client):
+        """
+        Test for profile_rule_add_remove when client cannot obtain a profile
+
+        Assert that the system exits
```

Comment:
We can fix these up when we next work on the files.  I'll fix up the ones I see.